### ISSUE TITLE
recode type promotion and performance issue

### ISSUE THREAD

**Issue by tk3369** [JuliaData/CategoricalArrays.jl#284]
The `recode` function does not properly recompute the eltype of the array.  In the following example, I expect `Union{Missing,Int64}` rather than `Union{Missing,Nothing,Int64}`:

```julia
julia> rs = [1, 2, nothing]
3-element Array{Union{Nothing, Int64},1}:
 1
 2
  nothing
julia> recode(rs, nothing => missing)
3-element Array{Union{Missing, Nothing, Int64},1}:
 1
 2
  missing
```

Also, it performs quite poorly as compared to regular array comprehension:

```julia
julia> x = rand([nothing,1,2,3,4,5], 10000);

julia> @btime recode($x, nothing => missing);
  219.951 μs (2 allocations: 88.02 KiB)

julia> n2m(A) = [x === nothing ? missing : x for x in A]

julia> @btime n2m($x);
  30.887 μs (5 allocations: 166.23 KiB)
```

It seems that this try-catch block is causing the slowness.  Do we really need it here?

https://github.com/JuliaData/CategoricalArrays.jl/blob/cfbca7431e5b15ba93733a46aae8620c4afb89ea/src/recode.jl#L63-L70

**Comment by bkamins** [comment]
And the first part of the issue is the assumption:
```
# Exception 1: if T === Missing and default not missing,
# assume the caller wants to recode only some values to missing,
# but accept original values
if T === Missing && !isa(default, Missing)
    dest = Array{Union{eltype(a), Missing}}(undef, size(a))
```
so
```
julia> recode(Union{String,Int}[1,"1"], 1 => missing)
2-element Array{Union{Missing, Int64, String},1}:
 missing
 "1"
```

**Comment by nalimilan** [comment]
Yes I wrote that code before `replace` in Base, where I applied a more general strategy of getting rid of all singleton types which are being replaced. Would you feel like adapting the `replace` code from here?
https://github.com/JuliaLang/julia/blob/b0a55b864106e6e82724e67de1b68d5db010ea4b/base/set.jl#L547

Regarding performance, maybe the `try` can be hoisted out of the loop? I think the custom error is useful as `recode` doesn't preserve the input type (contrary to `replace`), which is most often useful but potentially confusing.

**Comment by bkamins** [comment]
> Though are you sure the slowness is due to `try... catch`?

At least I have not checked this. It would have to be benchmarked.

**Comment by tk3369** [comment]
I actually checked last night.  Removing the try/catch totally remove the performance issue.
